flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExecutionJobVertexTaskInputsOutputsDescriptorBuilder.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.runtime.executiongraph.ExecutionJobVertex;
+import org.apache.flink.runtime.executiongraph.IntermediateResult;
+import org.apache.flink.runtime.jobgraph.DistributionPattern;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSet;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobgraph.JobEdge;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.shuffle.TaskInputsOutputsDescriptor;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Utils to build a {@link TaskInputsOutputsDescriptor} from an {@link ExecutionJobVertex}.
+ * The result descriptor would suites for an instance with max possible input channels and result subpartitions.
+ */
+final class ExecutionJobVertexTaskInputsOutputsDescriptorBuilder {
+
+	static TaskInputsOutputsDescriptor buildTaskInputsOutputsDescriptor(
+			final ExecutionJobVertex executionJobVertex,
+			final Map<JobVertexID, ExecutionJobVertex> vertices) {
+
+		final Map<IntermediateDataSetID, Integer> maxPossibleNumbersOfInputChannelsPerInstance =
+			getMaxPossibleNumbersOfInputChannelsPerInstance(executionJobVertex);
+		final Map<IntermediateDataSetID, Integer> maxPossibleNumbersOfResultSubpartitionsPerInstance =
+			getMaxPossibleNumbersOfResultSubpartitionsPerInstance(executionJobVertex, vertices);
+
+		return TaskInputsOutputsDescriptor.from(
+			maxPossibleNumbersOfInputChannelsPerInstance,
+			maxPossibleNumbersOfResultSubpartitionsPerInstance);
+	}
+
+	private static Map<IntermediateDataSetID, Integer> getMaxPossibleNumbersOfInputChannelsPerInstance(
+			final ExecutionJobVertex executionJobVertex) {
+
+		Map<IntermediateDataSetID, Integer> maxPossibleNumbers = new HashMap<>();
+		final List<JobEdge> inputEdges = executionJobVertex.getJobVertex().getInputs();
+		for (int i = 0; i < inputEdges.size(); i++) {
+			final JobEdge inputEdge =  inputEdges.get(i);
+			final IntermediateResult consumedResult = executionJobVertex.getInputs().get(i);
+
+			// the inputs order should match in JobGraph and ExecutionGraph
+			checkState(consumedResult.getId().equals(inputEdge.getSourceId()));
+
+			final int maxPossibleNumber = getMaxPossibleNumberOfEdgesToTarget(
+				executionJobVertex.getParallelism(),
+				consumedResult.getNumberOfAssignedPartitions(),
+				inputEdge.getDistributionPattern());
+			maxPossibleNumbers.put(consumedResult.getId(), maxPossibleNumber);
+		}
+
+		return maxPossibleNumbers;
+	}
```

Comment:
Good catch! Agreed that it is a problem in certain cases.
However, we can hardly resolve it at the moment since runtime does not know such behaviors of operators.
An alternative is to add a switch config "slot.allocation.respect-shuffle-memory". But FLIP-56 may also need to take it into consideration, otherwise a dynamic slot would acquire no network buffer if the switch is off.